pkg/apis/apps/v1alpha1/daemonset_types.go

```diff
@@ -0,0 +1,270 @@
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// DaemonSetUpdateStrategy is a struct used to control the update strategy for a DaemonSet.
+type DaemonSetUpdateStrategy struct {
+	// Type of daemon set update. Can be "RollingUpdate" or "OnDelete". Default is RollingUpdate.
+	// +optional
+	Type DaemonSetUpdateStrategyType `json:"type,omitempty" protobuf:"bytes,1,opt,name=type"`
+
+	// Rolling update config params. Present only if type = "RollingUpdate".
+	// +optional
+	RollingUpdate *RollingUpdateDaemonSet `json:"rollingUpdate,omitempty" protobuf:"bytes,2,opt,name=rollingUpdate"`
+}
+
+type DaemonSetUpdateStrategyType string
+type RollingUpdateType string
+
+const (
+	// Replace the old daemons by new ones using rolling update i.e replace them on each node one after the other.
+	RollingUpdateDaemonSetStrategyType DaemonSetUpdateStrategyType = "RollingUpdate"
+
+	// Replace the old daemons only when it's killed
+	OnDeleteDaemonSetStrategyType DaemonSetUpdateStrategyType = "OnDelete"
+
+	// StandardRollingUpdateType replace the old daemons by new ones using rolling update i.e replace them on each node one after the other.
+	// this is the default type for RollingUpdate.
+	StandardRollingUpdateType RollingUpdateType = "Standard"
+
+	// Replace container image without killing the pod.
+	InplaceRollingUpdateType RollingUpdateType = "Inplace"
+
+	// SurgingRollingUpdateType replaces the old daemons by new ones using rolling update i.e replace them on each node one
+	// after the other, creating the new pod and then killing the old one.
+	SurgingRollingUpdateType RollingUpdateType = "Surging"
+)
+
+// Spec to control the desired behavior of daemon set rolling update.
+type RollingUpdateDaemonSet struct {
+	// Type is to specify which kind of rollingUpdate.
+	Type RollingUpdateType `json:"rollingUpdateType,omitempty" protobuf:"bytes,1,opt,name=rollingUpdateType"`
+
+	// The maximum number of DaemonSet pods that can be unavailable during the
+	// update. Value can be an absolute number (ex: 5) or a percentage of total
+	// number of DaemonSet pods at the start of the update (ex: 10%). Absolute
+	// number is calculated from percentage by rounding up.
+	// This cannot be 0.
+	// Default value is 1.
+	// Example: when this is set to 30%, at most 30% of the total number of nodes
+	// that should be running the daemon pod (i.e. status.desiredNumberScheduled)
+	// can have their pods stopped for an update at any given
+	// time. The update starts by stopping at most 30% of those DaemonSet pods
+	// and then brings up new DaemonSet pods in their place. Once the new pods
+	// are available, it then proceeds onto other DaemonSet pods, thus ensuring
+	// that at least 70% of original number of DaemonSet pods are available at
+	// all times during the update.
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
```

Comment:
May be we can add a comment say that MaxUnavailable does not take effect if StandardRollingUpdateType is specified.

Reply:
Yes, in fact , `MaxUnavailable` only work in the cases that  RollingUpdateType is "StandardRollingUpdateType" or  "InplaceRollingUpdateType", not work for SurgingRollingUpdateType.